File: backend/models/model_config.py
"""
Pydantic models for LLM model configuration.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional
from datetime import datetime

from pydantic import BaseModel, Field, field_validator


def _clean_optional(value: Optional[str]) -> Optional[str]:
    if value is None:
        return None
    cleaned = value.strip()
    return cleaned if cleaned else None


class ModelConfigBase(BaseModel):
    """Shared fields for model configuration."""

    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    provider: str = Field(..., min_length=1, max_length=50)
    model: str = Field(..., min_length=1, max_length=100)
    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    max_tokens: Optional[int] = Field(default=None, ge=1)
    top_p: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    frequency_penalty: Optional[float] = Field(default=None, ge=-2.0, le=2.0)
    presence_penalty: Optional[float] = Field(default=None, ge=-2.0, le=2.0)
    api_key: Optional[str] = Field(default=None, min_length=1, max_length=255)
    base_url: Optional[str] = Field(default=None, min_length=1, max_length=255)
    extra_params: Dict[str, Any] = Field(default_factory=dict)

    @field_validator("name", "provider", "model", mode="before")
    @classmethod
    def strip_required_strings(cls, value):
        if not isinstance(value, str):
            raise ValueError("Must be a string")
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field cannot be empty")
        return cleaned

    @field_validator("description", "api_key", "base_url", mode="before")
    @classmethod
    def strip_optional_strings(cls, value):
        return _clean_optional(value)


class ModelConfigCreate(ModelConfigBase):
    """Create request for model configuration."""


class ModelConfigUpdate(BaseModel):
    """Update request for model configuration."""

    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    description: Optional[str] = Field(default=None, max_length=500)
    provider: Optional[str] = Field(default=None, min_length=1, max_length=50)
    model: Optional[str] = Field(default=None, min_length=1, max_length=100)
    temperature: Optional[float] = Field(default=None, ge=0.0, le=2.0)
    max_tokens: Optional[int] = Field(default=None, ge=1)
    top_p: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    frequency_penalty: Optional[float] = Field(default=None, ge=-2.0, le=2.0)
    presence_penalty: Optional[float] = Field(default=None, ge=-2.0, le=2.0)
    api_key: Optional[str] = Field(default=None, min_length=1, max_length=255)
    base_url: Optional[str] = Field(default=None, min_length=1, max_length=255)
    extra_params: Optional[Dict[str, Any]] = None
    version: Optional[int] = Field(default=None, ge=1)

    @field_validator("name", "provider", "model", mode="before")
    @classmethod
    def strip_optional_required_strings(cls, value):
        if value is None:
            return value
        if not isinstance(value, str):
            raise ValueError("Must be a string")
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field cannot be empty")
        return cleaned

    @field_validator("description", "api_key", "base_url", mode="before")
    @classmethod
    def strip_optional_strings(cls, value):
        return _clean_optional(value)


class ModelConfigRead(ModelConfigBase):
    """Response model for model configuration."""

    id: str
    version: int
    created_at: str
    updated_at: str


class ModelConfigImportItem(ModelConfigBase):
    """Item for import payload."""

    id: Optional[str] = None
    version: Optional[int] = Field(default=None, ge=1)
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class ModelConfigExport(BaseModel):
    """Export payload for model configs."""

    version: str = "1.0.0"
    exported_at: str = Field(default_factory=lambda: datetime.utcnow().isoformat())
    configs: List[ModelConfigRead]


class ModelConfigImport(BaseModel):
    """Import payload for model configs."""

    version: str = "1.0.0"
    configs: List[ModelConfigImportItem]
    overwrite: bool = False

    @field_validator("version")
    @classmethod
    def validate_version(cls, value: str) -> str:
        if value != "1.0.0":
            raise ValueError("Unsupported version. Expected 1.0.0")
        return value


class ModelConfigTestRequest(BaseModel):
    """Request body for testing a model configuration."""

    prompt: str = Field(default="Respond with the word 'pong'.", min_length=1, max_length=500)


class ModelConfigTestResponse(BaseModel):
    """Response for model configuration test."""

    success: bool
    message: str
    response: Optional[str] = None
    duration_ms: Optional[int] = None

File: frontend/src/types/modelConfig.ts
export type ModelConfigExtraParams = Record<string, unknown>;

export interface ModelConfigBase {
    name: string;
    description?: string;
    provider: string;
    model: string;
    temperature: number;
    max_tokens?: number | null;
    top_p?: number | null;
    frequency_penalty?: number | null;
    presence_penalty?: number | null;
    api_key?: string | null;
    base_url?: string | null;
    extra_params: ModelConfigExtraParams;
}

export interface ModelConfigCreate extends ModelConfigBase {}

export interface ModelConfigUpdate {
    name?: string;
    description?: string;
    provider?: string;
    model?: string;
    temperature?: number;
    max_tokens?: number | null;
    top_p?: number | null;
    frequency_penalty?: number | null;
    presence_penalty?: number | null;
    api_key?: string | null;
    base_url?: string | null;
    extra_params?: ModelConfigExtraParams | null;
    version?: number;
}

export interface ModelConfig extends ModelConfigBase {
    id: string;
    version: number;
    created_at: string;
    updated_at: string;
}

export interface ModelConfigExport {
    version: string;
    exported_at: string;
    configs: ModelConfig[];
}

export interface ModelConfigImportItem extends ModelConfigBase {
    id?: string | null;
    version?: number | null;
    created_at?: string | null;
    updated_at?: string | null;
}

export interface ModelConfigImport {
    version: string;
    configs: ModelConfigImportItem[];
    overwrite?: boolean;
}

export interface ModelConfigImportResult {
    created: number;
    updated: number;
    skipped: number;
}

export interface ModelConfigTestRequest {
    prompt?: string;
}

export interface ModelConfigTestResponse {
    success: boolean;
    message: string;
    response?: string | null;
    duration_ms?: number | null;
}
